pkg/fsm/metrics/internal/processing_duration.go

```diff
@@ -0,0 +1,159 @@
+package internal
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/btree"
+)
+
+type ProcessingStartTimes struct {
+	m          sync.Mutex
+	startTimes *btree.BTreeG[requestStartTime]
+}
+
+// requestStartTime is a wrapper around reconcile.Request that adds a Generation field.
+type requestStartTime struct {
+	// key is (namespace, name, generation)
+	Namespace  string
+	Name       string
+	Generation int64
+	// value is (start time, failed)
+	Time time.Time
+	// Failed indicates whether this request has been processed with a failure
+	Failed bool
+}
+
+func (r requestStartTime) key() string {
+	return fmt.Sprintf("%s/%s/%d", r.Namespace, r.Name, r.Generation)
+}
+
+func less(req1, req2 requestStartTime) bool {
+	// compare namespace, name, and generation
+	return req1.key() < req2.key()
+}
+
+func NewProcessingStartTimes() *ProcessingStartTimes {
+	return &ProcessingStartTimes{
+		// NOTE: 32 strikes a balance between memory usage and performance (prefer shallow depth)
+		// order of 32 matches what controller-runtime uses, https://github.com/kubernetes-sigs/controller-runtime/blob/b6c5897febe566008678f765ec5a3a1ea04a123a/pkg/controller/priorityqueue/priorityqueue.go#L65
+		startTimes: btree.NewG[requestStartTime](32, less),
+	}
+}
+
+// GetRange returns the processing start times for all requests with name, namespace, and generation <= observedGeneration.
+func (p *ProcessingStartTimes) GetRange(name string, namespace string, observedGeneration int64, success bool) []time.Time {
+	p.m.Lock()
+	defer p.m.Unlock()
```

Comment:
since we're not mutating, perhaps we can avoid unnecessary lock contention by using a `RWLock` here

Reply:
Updated to a RWMutex and updated the relevant locking/unlocking logic